pkg/mosconfig/disk.go

```diff
@@ -0,0 +1,327 @@
+		if err != nil {
+			log.Infof("bootDisk: failed to read gptTable on disk %s: %v", disk, err)
+			continue
+		}
+
+		espNum := 0
+		for n, p := range gptTable.Partitions {
+			if p.IsEmpty() {
+				continue
+			}
+			if espNum == 0 && isGptEsp(p) {
+				espNum = n + 1
+			}
+		}
+
+		if espNum != 0 {
+			return found(disk, espNum)
+		} else {
+			log.Debugf("%s: no boot information found", disk)
+			continue
+		}
+
+		// Here we had a bbp or a esp but no /boot.
+		log.Errorf("Disk %s had GPT with ESP part %d\n",
+			disk, espNum)
+	}
+
+	return BootDisk{}, fmt.Errorf("Did not find boot disks searched %s", strings.Join(candidates, ", "))
+}
+
+type BootDisk struct {
+	Disk string // The disk, e.g. /dev/sda
+	Esp  string // ESP partition, "" if none
+}
+
+// BootDiskInfo returns full dev paths, eg /dev/sda and /dev/sda1
+//
+// * bootdisk (where grub needs installing to)
+//   - must be GPT, the value should be the disk with a BBP partition.
+//
+// * efiPart - well defined efi partition (type is well known id).
+//   - finding this is straight forward. by ID.
+func BootDiskInfo() (BootDisk, error) {
+	candidates, err := fetchDiskCandidates()
+	if err != nil {
+		return BootDisk{}, err
+	}
+	return bootDiskInfo(candidates...)
+}
+
+func EspBootPartition() (string, string, error) {
+	bi, err := BootDiskInfo()
+	if err != nil {
+		return "", "", err
+	}
+	return bi.Disk, bi.Esp, err
+}
+
+func efiClearBootEntries() error {
+	args := []string{"efibootmgr", "-v"}
+	stdout, stderr, rc := RunCommandWithOutputErrorRc(args...)
+	if rc != 0 {
+		return errors.Errorf("Error querying existing boot entries:\nstdout: %s\nstderr: %s", stdout, stderr)
+	}
+	lines := strings.Split(string(stdout), "\n")
+	for _, l := range lines {
+		if !strings.HasPrefix(l, "Boot0") {
+			continue
+		}
+		num, err := strconv.ParseInt(l[4:8], 16, 64)
+		if err != nil {
+			log.Warnf("Failed to parse boot entry: %s", l)
+			continue
+		}
+		log.Infof("Removing boot entry %d (%s)", num, l)
+		err = RunCommand("efibootmgr", "--bootnum", l[4:8], "--delete-bootnum")
+		if err != nil {
+			log.Warnf("(Probably ok) Error removing boot entry: %s", l)
+		}
+	}
+
+	const enoent = "No such file or directory"
+	err := RunCommand("efibootmgr", "--delete-bootnext")
+	if err != nil && err.Error() != enoent {
+		log.Warnf("(Probably ok) Error running efibootmgr delete-bootnext: '%s'", err.Error())
+	}
+	err = RunCommand("efibootmgr", "--delete-bootorder")
+	if err != nil && err.Error() != enoent {
+		log.Warnf("(Probably ok) Error running efibootmgr delete-bootorder: '%s'", err.Error())
+	}
+	return nil
+}
+
+func toUCS2Native(input string) ([]byte, error) {
+	e := utf16.Encode([]rune(input))
+	kname := make([]byte, 0, len(e)*2)
+	for _, n := range e {
+		kname = append(kname, byte(n&0xFF00>>8), byte(n&0x00FF))
+	}
+	kname = append(kname, byte(0))
+	kname = append(kname, byte(0))
+
+	return kname, nil
+}
+
+func toUCS2IConv(input string) ([]byte, error) {
```

Comment:
https://stackoverflow.com/questions/30556584/string-to-ucs-2 has some good info on doing this in golang .

Reply:
The great thing about the internet vs golang is you can find 20 examples of how to do what you want, and the 16th one you try may actually work :)